feat: add versionable repository (#40013)

permil-successive opened the pull request:
@nxvinay  @Prabal555 

https://redmine.successive.tech/issues/40013

permil-successive committed:
feat: 🎸 add checking of exsiting user in Auth Middleware

add fetchMe REST API in user module

✅ Closes: #40012

permil-successive committed:
refactor: 💡 beautify the code

permil-successive committed:
fix: 🐛 change module name and permissions

permil-successive committed:
fix: 🐛 fix all the request changes in PR#14

permil-successive committed:
fix: 🐛 fix unhanled error & date validation

permil-successive committed:
Merge branch 'develop' of https://github.com/permil-st/javascript-server into feature/40012

permil-successive committed:
feat: 🎸 add Versionable Repository

✅ Closes: #40013

Prabal555 commented on src/controllers/user/controller.ts:
```diff
@@ -19,39 +19,57 @@ class Controller {
 
     try {
-      sendResponse(await this.userRepository.create(req.body), res, 'data inserted');
+      const data = await this.userRepository.create(req.body, req.user.originalId);
```
destructure

Prabal555 commented on src/controllers/user/controller.ts:
```diff
@@ -19,39 +19,57 @@ class Controller {
 
     try {
-      sendResponse(await this.userRepository.list(req.body), res, 'data fetched');
+      const data = await this.userRepository.list(req.query.skip, req.query.limit);
```
destructure missing

Prabal555 commented on src/controllers/user/controller.ts:
```diff
@@ -19,39 +19,57 @@ class Controller {
+  update = async (req: IRequest, res: Response, next: NextFunction) => {
     try {
-      sendResponse(await this.userRepository.update(req.body.id, req.body), res, 'data updated');
+      const data = await this.userRepository.update(req.body.id, req.body.dataToUpdate, req.user.originalId);
```
destructure missing

Prabal555 commented on src/controllers/user/controller.ts:
```diff
@@ -19,39 +19,57 @@ class Controller {
+  delete = async (req: IRequest, res: Response, next: NextFunction) => {
     try {
-      sendResponse(await this.userRepository.delete(req.params.id), res, 'data deleted');
+      const data = await this.userRepository.delete(req.params.id, req.user.originalId);
```
destructure missing

Prabal555 commented on src/libs/routes/authMiddleware.ts:
```diff
@@ -22,13 +32,27 @@ export default (currentModule: string, permissionType: string) => (req: Request,
+      return next(error);
+    }
+
+    req.user = user; // assigning user to request for further use
```
better shifting to locals

Prabal555 commented on src/repositories/versionable/VersionableRepository.ts:
```diff
@@ -0,0 +1,101 @@
+export default class VersionableRepository<D extends IVersionableDocument, M extends mongoose.Model<D>> {
+  model: M;
+  _ERROR_CODE = 400;
+  private _deleteQuery = { deletedAt: {'$exists': false} };
```
why is _ prefixed with variable here

permil-successive replied:
I am used _ for showing constants. Now dropped.

Prabal555 commented on src/repositories/versionable/VersionableRepository.ts:
```diff
@@ -0,0 +1,101 @@
+    return mongoose.Types.ObjectId().toHexString();
+  }
+
+  private generateBy(currentUser: string) {
```
Generic method variable should not be named like 'user'

permil-successive replied:
this **currentUser** variable is used for that user who is performing operation on record. It is common for every record.

Prabal555 commented on src/repositories/versionable/VersionableRepository.ts:
```diff
@@ -0,0 +1,101 @@
+    return await this.model.countDocuments(query);
+  }
+
+  async create(data, currentUser: string): Promise<D> {
```
variable name should be as user

permil-successive committed:
fix: 🐛 add CRUD methods in User Repo, use destructing

remove IRequest file because using res.locals, implements all the
changes request in PR #15 #14

✅ Closes: #40013

Prabal555 commented:
Http Status code and request returns different code
![image](https://user-images.githubusercontent.com/45360184/74649531-c7b1e680-51a5-11ea-9018-9ffbe31bf5de.png)

Email is not validated properly in the data to update, Mobile number is not validated properly in the data to update and the dob is returning two errors in the same message.
![image](https://user-images.githubusercontent.com/45360184/74649914-8f5ed800-51a6-11ea-8d85-56937f5e00e6.png)

Error message is not correct for invalid hobbies type
![image](https://user-images.githubusercontent.com/45360184/74649885-8110bc00-51a6-11ea-90a1-4ec7fe38938f.png)

An incorrect response is returned from the service, try to **never return internal server error from the server** please make sure to check the validity of thing first like date in this case
![image](https://user-images.githubusercontent.com/45360184/74650065-e06ecc00-51a6-11ea-8623-5047bba880b8.png)

Prabal555 commented on src/repositories/versionable/VersionableRepository.ts:
```diff
@@ -0,0 +1,121 @@
+    this.isOperationSuccess(updatedData);
+
+    const updatedPreviousData = await this.MODEL.findByIdAndUpdate({ _id: originalData._id }, this.generateDeletedInfo(currentUser));
+    this.isOperationSuccess(updatedPreviousData);
```
what kind of variable name is updatedPreviousData

permil-successive replied:
Sir,

actually this variable stores the result of update query on original document for updating deleteAt. May you help me suggesting better name.

Prabal555 commented on src/repositories/versionable/VersionableRepository.ts:
```diff
@@ -0,0 +1,121 @@
+      ...originalData.toJSON(),
+      _id: this.generateId(),
+      ...this.generateUpdatedInfo(currentUser),
+      createdBy: this.generateCreatedInfo(currentUser).createdBy,
```
these methods could have been generally called getCreatedInfo instead of generatetCreatedInfo

Prabal555 commented:
Please make sure that created and updated at dates are same:

![image](https://user-images.githubusercontent.com/45360184/74650868-a1417a80-51a8-11ea-801f-17683efbc70b.png)

permil-successive commented:
> Please make sure that created and updated at dates are same:
> 
> ![image](https://user-images.githubusercontent.com/45360184/74650868-a1417a80-51a8-11ea-801f-17683efbc70b.png)

done

permil-successive committed:
fix: 🐛 changes as requested in PR#15

fix error handler for not picking proper error code, add validation for
dataToUpdate, add validation for email, add check for unique email in
user controller

✅ Closes: #40013

permil-successive commented:
> Http Status code and request returns different code
> ![image](https://user-images.githubusercontent.com/45360184/74649531-c7b1e680-51a5-11ea-9018-9ffbe31bf5de.png)
> 
> Email is not validated properly in the data to update, Mobile number is not validated properly in the data to update and the dob is returning two errors in the same message.
> ![image](https://user-images.githubusercontent.com/45360184/74649914-8f5ed800-51a6-11ea-8d85-56937f5e00e6.png)
> 
> Error message is not correct for invalid hobbies type
> ![image](https://user-images.githubusercontent.com/45360184/74649885-8110bc00-51a6-11ea-90a1-4ec7fe38938f.png)
> 
> An incorrect response is returned from the service, try to **never return internal server error from the server** please make sure to check the validity of thing first like date in this case
> ![image](https://user-images.githubusercontent.com/45360184/74650065-e06ecc00-51a6-11ea-8623-5047bba880b8.png)

resolved
